### ISSUE TITLE
Opportunity for optimisation when querying by pk

### ISSUE THREAD

**Issue by hrj** [hasura/graphql-engine#827]
I love the addition of the "analyze" button in the latest release; thanks!

For the following graphql query:
```
query ws {
  workspace_by_pk(id:1){
    name
  }
}
```

The SQL generated is:
```
SELECT
  coalesce((json_agg("root") -> 0), 'null') AS "root"
FROM
  (
    SELECT
      row_to_json(
        (
          SELECT
            "_1_e"
          FROM
            (
              SELECT
                "_0_root.base"."name" AS "name"
            ) AS "_1_e"
        )
      ) AS "root"
    FROM
      (
        SELECT
          *
        FROM
          "public"."workspace"
        WHERE
          (
            (("public"."workspace"."id") = ('1'))
            OR (
              (("public"."workspace"."id") IS NULL)
              AND (('1') IS NULL)
            )
          )
      ) AS "_0_root.base"
  ) AS "_2_root_alias"
```

I see two opportunities for simplification:
* The clause for NULL check seems redundant when the pk value is known to be not NULL.
* The nested `select` clauses. The query generator could coalesce them before sending them to the DB.

I don't know whether the above will actually improve performance for this simple query. Also, my dataset is small. However, eliminating the extra clauses might benefit other complex queries, and especially so on larger datasets.

**Comment by 0x777** [comment]
Glad that you like it ! You can expect a lot more improvements in the coming few days. 

> The clause for NULL check seems redundant when the pk value is known to be not NULL.

True. This is redundant. Can be removed. 

> The nested select clauses. The query generator could coalesce them before sending them to the DB.

We've spent a fair bit of time to see if these extra nesting of selects has any performance impact. It doesn't affect Postgres at all, even in complex queries.

**Comment by 0x777** [comment]
> The clause for NULL check seems redundant when the pk value is known to be not NULL.
Fixed in ae63ed96033502e721de5dba8dcc75b56249b1f7
